app/src/main/java/me/guillaumin/android/osmtracker/gpx/ExportAndShareTask.java

```diff
@@ -0,0 +1,168 @@
+import android.content.pm.PackageManager;
+import android.database.Cursor;
+import android.net.Uri;
+import android.util.Log;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Date;
+import java.util.TreeMap;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import me.guillaumin.android.osmtracker.R;
+import me.guillaumin.android.osmtracker.exception.ExportTrackException;
+
+/**
+ * Exports the track and media (if any) to a a temporary file, then shares it with another app
+ */
+
+public class ExportAndShareTask extends ExportTrackTask {
+    private static final String TAG = ExportToTempFileTask.class.getSimpleName();
+
+    /**
+     * Maps from name that is passed in URI to another app to a real file in the cache dir.
+     */
+    static final TreeMap<String, File> sharedFiles = new TreeMap<String, File>();
+    /**
+     * Name of the file, passed to another app in URI.
+     */
+    private String filename;
+    /**
+     * Subfolder in cacheDir, where track is exported
+     */
+    private File trackDir;
+
+    public ExportAndShareTask(Context context, long trackId) {
+        super(context, trackId);
+        File cacheDir = context.getCacheDir();
+        // Generate name for a temporary directory
+        trackDir = new File(cacheDir, "trackDirectory");
+        int i = 0;
+        while (trackDir.exists())
+            trackDir = new File(cacheDir, "trackDirectory" + i++);
+        if (trackDir.mkdir())
+            Log.d(TAG, "Temporary file: " + trackDir.getAbsolutePath());
+        else {
+            Log.e(TAG, "Could not create temporary directory");
+            throw new IllegalStateException("Could not create temporary directory");
+        }
+    }
+
+    /**
+     * @param startDate
+     * @return The directory in which the track file should be created
+     * @throws ExportTrackException
+     */
+    @Override
+    protected File getExportDirectory(Date startDate) throws ExportTrackException {
+        return trackDir;
+    }
+
+    /**
+     * Whereas to export the media files or not
+     *
+     * @return
+     */
+    @Override
+    protected boolean exportMediaFiles() {
+        return true;
+    }
+
+    @Override
+    protected String buildGPXFilename(Cursor c) {
+        filename = super.buildGPXFilename(c);
+        return filename;
+    }
+
+    /**
+     * Whereas to update the track export date in the database at the end or not
+     *
+     * @return
+     */
+    @Override
+    protected boolean updateExportDate() {
+        return true;
+    }
+
+    @Override
+    protected void onPostExecute(Boolean success) {
+        super.onPostExecute(success);
+
+        File cacheDir = context.getCacheDir();
+        File fileToShare;
+        File[] generatedFiles = trackDir.listFiles();
+        boolean a;
```

Comment:
Actually I've added that variable for debugging purpose only (thus the meaningless name), but forgot to remove it.